pageserver/src/tenant/timeline.rs

```diff
@@ -2680,52 +2680,57 @@ impl Timeline {
         cutoff_horizon: Lsn,
         pitr: Duration,
     ) -> anyhow::Result<()> {
-        let mut gc_info = self.gc_info.write().unwrap();
-
-        gc_info.horizon_cutoff = cutoff_horizon;
-        gc_info.retain_lsns = retain_lsns;
-
-        // Calculate pitr cutoff point.
-        // If we cannot determine a cutoff LSN, be conservative and don't GC anything.
-        let mut pitr_cutoff_lsn: Lsn;
-
-        if pitr != Duration::ZERO {
-            // conservative, safe default is to remove nothing, when we have no
-            // commit timestamp data available
-            pitr_cutoff_lsn = *self.get_latest_gc_cutoff_lsn();
-
-            // First, calculate pitr_cutoff_timestamp and then convert it to LSN.
-            // If we don't have enough data to convert to LSN,
-            // play safe and don't remove any layers.
+        // First, calculate pitr_cutoff_timestamp and then convert it to LSN.
+        //
+        // Some unit tests depend on garbage-collection working even when
+        // CLOG data is missing, so that find_lsn_for_timestamp() doesn't
+        // work, so avoid calling it altogether if time-based retention is not
+        // configured. It would be pointless anyway.
+        let ts_cutoff: Option<LsnForTimestamp>;
+        ts_cutoff = if pitr != Duration::ZERO {
             let now = SystemTime::now();
             if let Some(pitr_cutoff_timestamp) = now.checked_sub(pitr) {
                 let pitr_timestamp = to_pg_timestamp(pitr_cutoff_timestamp);
 
-                match self
-                    .find_lsn_for_timestamp(pitr_timestamp)
-                    .no_ondemand_download()?
-                {
-                    LsnForTimestamp::Present(lsn) => pitr_cutoff_lsn = lsn,
-                    LsnForTimestamp::Future(lsn) => {
-                        debug!("future({})", lsn);
-                        pitr_cutoff_lsn = gc_info.horizon_cutoff;
-                    }
-                    LsnForTimestamp::Past(lsn) => {
-                        debug!("past({})", lsn);
-                    }
-                    LsnForTimestamp::NoData(lsn) => {
-                        debug!("nodata({})", lsn);
-                    }
-                }
-                debug!("pitr_cutoff_lsn = {:?}", pitr_cutoff_lsn)
+                Some(
+                    self.find_lsn_for_timestamp(pitr_timestamp)
+                        .no_ondemand_download()?,
```

Comment:
This is `async`, now lock-less (thanks to this PR) context that is able to do on-demand downloads.
There was a case when info update required layer downloads, so we should make it do so.

> no_ondemand_download()?

We should replace it with something like https://github.com/neondatabase/neon/pull/3199/files#diff-74152cb7b3681c454c4156d530ce544ee8f15048c9f3cacc736dba5dc7c0d664R2701

that essentially triggers `with_ondemand_download(|| self.find_lsn_for_timestamp(pitr_timestamp)).await?`.

Reply:
Changed that. I'm going to keep that change as a separate commit for clarity.